keps/349-avoid-coscheduling/README.md

```diff
@@ -0,0 +1,418 @@
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- guarantee that two workloads would not schedule pods concurrently. Example
+scenarios in which two workloads may still concurrently schedule their pods:
+  - when succeeded pods are replaced with new because job's parallelism is less than its completions;
+  - when a failed pod gets replaced
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+We introduce a mechanism to avoid concurrent scheduling of Job pods. More
+precisely, we block admission of new workloads until the first batch of the
+Job is scheduled. Additionally, we await with starting the Job until all other
+admitted workloads have their pods scheduled. This behavior can be opted-in by
+a Job by a new annotation.
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a machine learning researcher I want to run a workload, comprising tens of
+pods which need to all run at the same time. Autoscaling the cluster and
+scheduling of all pods in this workload take a couple of minutes and I want to
+have a guarantee that other workloads scheduled within this time interval would
+not delay execution of my workload.
+
+My use can be supported by adding `kueue.x-k8s.io/wait-for-pods-scheduled`
+annotation to the jobs I run.
+
+#### Story 2
+
+As an administrator of a batch processing in my organization I want to be able
+to configure kueue quotas for on-demand resources a little higher than what is
+available in the quiet period. However, I don't want to risk two workloads
+deadlock in case of issues with provisioning nodes to satisfy the configured
+quota.
+
+My use can be supported by adding `kueue.x-k8s.io/wait-for-pods-scheduled`
+annotation to all jobs.
+
+### Notes/Contraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+If the workload fails to schedule its pods it could block admission of other
+workloads indefinitely.
+
+The risk can be mitigated by using the Job's `ActiveDeadlineSeconds` timeout
```

Comment:
It's really a serious problem, when other types of jobs do not support this feature. Even for native batch Job, pods can't run successfully for various of reasons besides resource problems.

Reply:
Agree with the sentiment, was just considering decoupling it from this proposal  as it is already an issue to some extent. Still, moved it to Design Details as the need was expressed by reviewers.